go/01_arrays_and_hashing/665_Non_decreasing_Array/238_Product_of_Array_Except_Self/238_Product_of_Array_Except_Self.go

```diff
@@ -0,0 +1,60 @@
+package productExceptSelf
+
+func getPrefixProduct(nums []int) []int {
+	prefixProductArr := make([]int, len(nums))
+	prefixProduct := 1
+	for i := range nums {
+		prefixProductArr[i] = prefixProduct * nums[i]
+		prefixProduct = prefixProductArr[i]
+	}
+	return prefixProductArr
+}
+
+func getPostFixProduct(nums []int) []int {
+	postFixProductArr := make([]int, len(nums))
+	postFixProduct := 1
+	i := len(nums) - 1
+	for i >= 0 {
+		postFixProductArr[i] = postFixProduct * nums[i]
+		postFixProduct = postFixProductArr[i]
+		i--
+	}
+	return postFixProductArr
+}
+
+func productExceptSelf(nums []int) []int {
+	result := make([]int, len(nums))
+	prefixProductArr := getPrefixProduct(nums)
+	postFixProductArr := getPostFixProduct(nums)
+	prefixValue := 1
+	postFixValue := 1
+
+	for i := range result {
```

Comment:
I think these if statements could be avoided, for example:
```go
func productExceptSelf(nums []int) []int {
    res := make([]int, len(nums))
    
    prefix := 1
    res[0] = 1
    for i := 0; i < len(nums); i++ {
        res[i] = prefix
        
        prefix *= nums[i]
    }
    
    postfix := 1
    for i := len(nums)-1; i >= 0; i-- {
        res[i] *= postfix
        
        postfix *= nums[i]
    }
    
    return res
}
```

Reply:
Yes agreed. The solution I have written is a suboptimal one where I consume extra space for prefix and postfix array.

The proposed solution uses O(1) space and a cleaner way of doing things. I will add it as a new procedure and write tests. Thanks for the suggestion